### PR TITLE
[BEAM-147] Adding proto buff matchers for dataflow client

### PR DESCRIPTION
**Pull request by pabloem**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
Adding these matchers to test changes to dataflow client - service communication in metrics.

### CODE REVIEW

**Comment by pabloem** [comment]
r: @charlesccychen 
I'm adding these matchers to test changes to dataflow client - service communication in metrics.

File: sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+from hamcrest.core.base_matcher import BaseMatcher
+
+import apache_beam.internal.clients.dataflow as dataflow
+
+
+IGNORED = object()
+
+
+class MetricStructuredNameMatcher(BaseMatcher):
+  """ Matches a MetricStructuredName."""
```
Comment (charlesccychen) [inline sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py]: Please omit spaces at beginning of docstring

File: sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    self.origin = origin
+    self.context = context
+
+  def _matches(self, item):
+    if self.name != IGNORED and item.name != self.name:
+      return False
+    if self.origin != IGNORED and item.origin != self.origin:
+      return False
+    if self.context != IGNORED:
+      for key, name in self.context.items():
```
Comment (charlesccychen) [inline sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py]: best-practice is to call iteritems() when iterating like this.

File: sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        if key != IGNORED and key not in item.context:
```
Comment (charlesccychen) [inline sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py]: when the user ever set "key" to IGNORED?

File: sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  def describe_to(self, description):
+    descriptors = []
+    if self.name != IGNORED:
+      descriptors.append('name is {}'.format(self.name))
+    if self.origin != IGNORED:
+      descriptors.append('origin is {}'.format(self.origin))
+    if self.context != IGNORED:
+      descriptors.append('context is ({})'.format(str(self.context)))
+
+    item_description = '{}'.format(' and '.join(descriptors))
```
Comment (charlesccychen) [inline sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py]: outer "'{}'.format(" seems unnecessary

File: sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  def describe_to(self, description):
+    descriptors = []
+    if self.cumulative != IGNORED:
+      descriptors.append('cumulative is {}'.format(self.cumulative))
+    if self.name != IGNORED:
+      descriptors.append('name is {}'.format(self.name))
+    if self.scalar != IGNORED:
+      descriptors.append('scalar is ({})'.format(str(self.scalar)))
+
+    item_description = '{}'.format(' and '.join(descriptors))
```
Comment (charlesccychen) [inline sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py]: outer "'{}'.format(" seems unnecessary

**Comment by charlesccychen** [comment]
Thanks!  Just some minor comments.

File: sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py (hunk 1 above)
Comment (pabloem) [inline sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py]: Oops yes.

File: sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py (hunk 2 above)
Comment (pabloem) [inline sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py]: Done.

File: sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py (hunk 3 above)
Comment (pabloem) [inline sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py]: humm you are correct, this does not make sense. I'll remove 'wildcards' from this section.

File: sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py (hunk 5 above)
Comment (pabloem) [inline sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py]: You're right. I don't know why I did this. : )

File: sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py (hunk 4 above)
Comment (pabloem) [inline sdks/python/apache_beam/internal/clients/dataflow/message_matchers.py]: You're right. I don't know why I did this. : )

**Comment by charlesccychen** [comment]
Thanks! LGTM

R: @robertwb for merge

**Comment by charlesccychen** [comment]
Actually, also note that the commit message should spell abbreviate "protocol buffers" as "protobuf".

**Comment by pabloem** [comment]
Fixed the commit message and squashed. Thanks Charles.

**Pull request merged**